internal/transport/transport_test.go

```diff
@@ -1663,81 +1663,279 @@ func (s) TestReadGivesSameErrorAfterAnyErrorOccurs(t *testing.T) {
+		}
+	}{
+		// Note: multiple authority headers are handled by the framer itself,
+		// which will cause a stream error. Thus, it will never get to
+		// operateHeaders with the check in operateHeaders for possible grpc-status sent back.
+
+		// multiple :authority or multiple Host headers would make the eventual
+		// :authority ambiguous as per A41. This takes precedence even over the
+		// fact a request is non grpc. All of these requests should be rejected
+		// with grpc-status Internal.
+		{
+			name: "Multiple host headers non grpc",
+			headers: []struct {
+				name   string
+				values []string
+			}{
+				{name: ":method", values: []string{"POST"}},
+				{name: ":path", values: []string{"foo"}},
+				{name: ":authority", values: []string{"localhost"}},
+				{name: "host", values: []string{"localhost", "localhost2"}},
+			},
+		},
+		{
+			name: "Multiple host headers grpc",
+			headers: []struct {
+				name   string
+				values []string
+			}{
+				{name: ":method", values: []string{"POST"}},
+				{name: ":path", values: []string{"foo"}},
+				{name: ":authority", values: []string{"localhost"}},
+				{name: "content-type", values: []string{"application/grpc"}},
+				{name: "host", values: []string{"localhost", "localhost2"}},
+			},
+		},
 	}
+	for _, test := range tests {
+		server := setUpServerOnly(t, 0, &ServerConfig{}, suspended)
+		defer server.stop()
+		// Create a client directly to not tie what you can send to API of
+		// http2_client.go (i.e. control headers being sent).
+		mconn, err := net.Dial("tcp", server.lis.Addr().String())
+		if err != nil {
+			t.Fatalf("Client failed to dial: %v", err)
+		}
+		defer mconn.Close()
 
-	// success chan indicates that reader received a RSTStream from server.
-	// An error will be passed on it if any other frame is received.
-	success := testutils.NewChannel()
+		if n, err := mconn.Write(clientPreface); err != nil || n != len(clientPreface) {
+			t.Fatalf("mconn.Write(clientPreface) = %d, %v, want %d, <nil>", n, err, len(clientPreface))
+		}
 
-	// Launch a reader goroutine.
-	go func() {
-		for {
-			frame, err := framer.ReadFrame()
-			if err != nil {
-				return
+		framer := http2.NewFramer(mconn, mconn)
+		framer.ReadMetaHeaders = hpack.NewDecoder(4096, nil)
+		if err := framer.WriteSettings(); err != nil {
+			t.Fatalf("Error while writing settings: %v", err)
+		}
+
+		// success chan indicates that reader received a Headers Frame with
+		// desired grpc status and message from server. An error will be passed
+		// on it if any other frame is received.
+		success := testutils.NewChannel()
+
+		// Launch a reader goroutine.
+		go func() {
+			for {
+				frame, err := framer.ReadFrame()
+				if err != nil {
+					return
+				}
+				switch frame := frame.(type) {
+				case *http2.SettingsFrame:
+					// Do nothing. A settings frame is expected from server preface.
+				case *http2.MetaHeadersFrame:
+					for _, header := range frame.Fields {
+						if header.Name == ":status" {
```

Comment:
This check will pass if header doesn't have `:status`.
Keep the value in a var, and do the checks after the for loop.